### ISSUE TITLE
[FEAT] On file reload : do **NOT** flush undo/redo

### ISSUE THREAD

**Issue by Kochise** [notepad-plus-plus/notepad-plus-plus#12766]
### Description of the Issue

File opened in editor gets modified by external program.
The file is updated in the editor (automatically or manually).
The previous version is lost forever, the undo/redo arrows are disabled. 

### Steps to Reproduce the Issue

1. See Description just above
2. 
3. 

### Expected Behavior

If file is reloaded due to an external change :
1. Store current file in undo history
2. Reload file
3. Flush redo

### Actual Behavior

1. See Description just above
2. 
3. 

### Debug Information

No debug to provide.

**Comment by alankilborn** [comment]
This was tried before and was so unsuccessful that it was reverted.  I doubt the developers are going to try it again.

**Comment by Kochise** [comment]
In what was it unsuccessful ?

**Comment by alankilborn** [comment]
Search old issues/change logs.

**Comment by rdipardo** [comment]
@Kochise,
> In what was it unsuccessful ?

Cf. <https://github.com/notepad-plus-plus/notepad-plus-plus/issues/5626#issuecomment-499858088>:

> the biggest problem was, that every reload re-populated the undo buffer, regardless if the file itself was modified or not, causing to run out of memory if either done frequently, or reloading huge files.
>
> see here: #5595
> 
> and here: https://notepad-plus-plus.org/community/topic/17530/performance-hit-due-to-allow-undo-on-reload/3

**Comment by Kochise** [comment]
Ok, could it be possible to enable it in `Settings.../MISC` ?

**Comment by alankilborn** [comment]
>Ok, could it be possible to enable it in Preferences.../MISC/File Status Auto-Detection ?

As I said, I don't think developers are going to revisit anything on this.

>was editing a self-modifying file, it emptied itself, lost everything

Meaning you got a prompt to reload and said yes to that prompt?

**Comment by Kochise** [comment]
Nope, no prompt, auto-reloaded itself, cleaned the undo buffer.

**Comment by alankilborn** [comment]
What's your setting here:

![image](https://user-images.githubusercontent.com/77065706/211191769-f943f3ed-74ee-4525-bb85-58eb5cb261ab.png)

**Comment by alankilborn** [comment]
Also, for developers to consider your issue, provide _Debug Information_.

**Comment by Kochise** [comment]
> What's your setting here:

That's how it is... NOW. Before the two boxes were checked.

**Comment by alankilborn** [comment]
Nobody cares how it is NOW; they care how it was when you encountered the conditions of your issue.

_Update silently_ is a bad choice if there is the possibility that you are doing editing of the file and there is the possibility it could be changed by some external process.  Perhaps this is obvious now.

**Comment by Kochise** [comment]
Perhaps. The settings are so "hard" to find that I didn't paid attention to it, until it was too late.

**Comment by alankilborn** [comment]
>The settings are so "hard" to find that I didn't paid attention to it, until it was too late.

Probably best to leave this one at the "default" value (as I showed in the screenshot), unless one really knows what they're doing.  I know this is after-the-fact of loss for you; sorry to hear that.

**Comment by Kochise** [comment]
No problem, life is like that. No one was hurt.

**Comment by NathanSweet** [comment]
Auto reload is good. Losing all undo because of it is very bad. There is no reason to do so, just put the old contents into the undo buffer and keep everything. This is an important issue to fix!

**Comment by Kochise** [comment]
It is said that you might run out of your GB of memory. Unless the undo buffer is given a size limit.

**Comment by NathanSweet** [comment]
That is unlikely for most editing. If it is a worry consider reasonable mitigation. Maybe disable for extremely large files. Multiple reloads with no other changes could use a single undo slot. Add lightweight compression (eg LZ4).

In general, separate from this issue, undo size has a limit and should probably have controls.

FWIW, other editors, like Eclipse, auto reload and push the file into an undo slot without any special configuration for that.

**Comment by mpheath** [comment]
@NathanSweet 

> Losing all undo because of it is very bad.

Agreed, though inserting a whole document into the Undo stack is not an optimal solution. Only the changes are needed to be inserted from the file to sync with the buffer.

If the buffer is in a modified state and the file is changed, then reloading may lose the modifications as it becomes a choice between file or buffer content with no other option. So some merging might be useful to avoid lose of the modified data.

Patching the buffer can help to keep the Undo state and update the buffer with just the changes. This is a PythonScript 3 script that may assist to [patch the buffer](https://gist.github.com/mpheath/8cb74b9fb13d362dd6c27e14d721815b) .

> FWIW, other editors, like Eclipse, auto reload and push the file into an undo slot without any special configuration for that.

SciTE does similar with keeping Undo based on the property setting `reload.preserves.undo`. Even though this convenience may exist in some editors, the modifications are something that I may not want lost to the Undo stack after a reload.

**Comment by NathanSweet** [comment]
True, maybe diff the contents being replaced and only push an undo state with the changes. If the diff is difficult though, I would skip it for v1.0 of this feature. RAM is plentiful these days and I _want_ to use it to keep my undo states. Excessive reloads could be collapsed into a single undo state, so I don't see this running away with memory, plus it could be disabled by file size.

I've never seen any editor merge modifications with a reload. If there are modifications and the file changes, the reload should prompt (as it does currently) to either reload and lose modifications or don't reload. No merging is necessary and I think outside the editor's scope. If you don't want to lose your modifications, you have the chance to not reload and save them (save a copy of the file, do a merge, etc).

There are a few challenges, but they can be worked around -- I am confident of this because many other editors do so. It's a very important feature to keep the undo state through reload.

**Comment by donho** [comment]
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12766#issuecomment-1374418025

**Comment by NathanSweet** [comment]
Re: comment #12766, there are many ways to workaround those issues, mentioned above.

**Comment by weirdalsuperfan** [comment]
I think this is a duplicate of https://github.com/notepad-plus-plus/notepad-plus-plus/issues/5813

A lot of people want this back as an optional setting. It seems it was removed due to 1 or 2 people complaining about it. Far more people have since requested it back.
